keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
+A newly introduced validation in api-server will reject deletion requests for resources with non-empty `Liens` and return ["409 Conflict" error code](https://datatracker.ietf.org/doc/html/rfc2616#section-10.4.10).
+
+`Liens` is defined as a slice of strings, like `Finalizers`.
+The strings need to be namespaced keys.
+Multiple users or controllers can add their `Liens` for their own purpose, and they can remove their own `Liens` when it is no longer needed.
+`Liens` should be added per controller or per user basis.
+Deletion requests of the resource are blocked until the last `Liens` for the resource is removed
+(Just to be clear, the difference between `Liens` and `Finalizers` is that `Liens` blocks the deletion request itself, while `Finalizers` blocks the deletion to be completed).
+
+A PoC implementation for in-use protection can be found, [here](https://github.com/mkimuram/kubernetes/commits/lien).
+Also, how it can be consumed by Secret protection controller can be found, [here](https://github.com/mkimuram/secret-protection/tree/lien).
+
+### API Changes
+
+`Liens` is added to `ObjectMeta`.
+
+```go
+Liens []string
+```
+
+Validation criteria of the field are as follows:
+- Keys must be namespaced (example: `kubernetes.io/secret-protection`;  `foo.example/bar`)
+- Maximum length of each key is 253 characters
+- Maximum number of keys is 32
+
+### Other Design Considerations
+
+#### Behavior with `ownerReference`
+
+Lien itself only blocks a deletion request to an object that is added to.
+However, it may not be clear how lien behaves when used with [`ownerReference`](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#owners-dependents).
+Therefore, this section describes the behavior.
+
+In summary, lien will block cascading deletion, but would not block deletion of any dependent resources individually.
+
+For example, let's assume that we have Deployment A which manages Replicaset B which manages Pod C.
+In this situation, ownerRefernces are set from Pod C to Replicaset B and from Replicaset B to Deployment A.
+
+If a lien is set to Deployment A, only a deletion request to Deployment A is blocked.
+Users can still request to delete Replicaset B and Pod C directly, but they can't request to delete them through Deployment A.
+
+If a lien is set to Replicaset B, only a deletion request to Replicaset B is blocked.
+Users can still request to delete Deployment A and Pod C directly.
+When deletion of the Deployment A is requested, whether it completes immediately or not depends on the cascading policy and Replicaset B's `blockOwnerDeletion`.
```

Comment:
Does this mean that the Lien is ignored and the resource is deleted when it's owner resource is deleted?  That would seem to imply that if you really need the Lien you should place it on the top-level resource and not on any of the lower-level resources.

Put another way, Liens are enforced by the API endpoint and NOT by the garbage collector.  So if an API Delete request is made for an object with a Lien, that gets enforced.  But if an object with a Lien is part of a set of ownerReferences and the top-level owner does not have a Lien, any Liens on the other resources in the graph are ignored.